src/main/java/com/google/sps/servlets/ServeImageServlet.java

```diff
@@ -0,0 +1,50 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.googl.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * When the user submits the form, Blobstore processes the file upload and then forwards the request
```

Comment:
What request is forwarded, from where? What exactly do you mean by forwarding a request? Can we find a way to make it easier to understand what's going on?

Actually, maybe there's a better place to put that information. In general, the documentation for Thing X should describe what Thing X does and how to use it. But it should avoid describing how Thing X is used by Thing Y, because it's likely that somebody will later change Thing Y without updating the documentation for Thing X.

Reply:
How does
Serves the blob by passing the blob key to blobstore service; replaces the getServingUrl method
sound? My understanding of what this class does is still a little lacking, this was the part that was the work around for the errors in the original walkthrough. (From this documentation https://cloud.google.com/appengine/docs/standard/java/blobstore#Serving_a_blob)